contracts/lottery/src/lib.rs

```diff
@@ -0,0 +1,608 @@
+type LotteryResult = Vec<u32>;
+
+#[contract]
+pub struct LotteryContract;
+
+#[contractimpl]
+impl LotteryContract {
+    /// It initializes the contract with all the needed parameters.
+    /// This function must be invoked byt the administrator just
+    /// after the contract deployment. 
+    /// It invokes the `create_lottery` function at the end of initialization.
+    ///
+    /// # Arguments
+    ///
+    /// - `env` - The environment for this contract.
+    /// - `admin` - Admin account address.
+    /// - `token` - The asset contract address we are using for this lottery. See [token interface](https://soroban.stellar.org/docs/reference/interfaces/token-interface).
+    /// - `ticket_price` - Unitary ticket price for the current lottery.
+    /// - `number_of_numbers` - Number of numbers possible to select by players
+    /// - `max_range` - Right boundary of the range players will select numbers from (1, max_range)
+    /// - `thresholds` - Thresholds with prizes for correctly selected numbers (specified as percentage of the pool balance)
+    /// - `min_players_count` - Minimum number of players needed to play the lottery
+    pub fn init(
+        env: Env,
+        admin: Address,
+        token: Address,
+        ticket_price: i128,
+        number_of_numbers: u32,
+        max_range: u32,
+        thresholds: Map<u32, u32>,
+        min_players_count: u32,
+    ) {
+        admin.require_auth();
+        let storage = env.storage().persistent();
+
+        if storage
+            .get::<_, LotteryState>(&DataKey::LotteryState)
+            .is_some() {
+            panic_with_error!(&env, Error::AlreadyInitialized);
+        }
+        storage.set(&DataKey::Admin, &admin);
+        storage.set(&DataKey::Token, &token);
+        storage.set(&DataKey::LotteryState, &LotteryState::Initialized);
+        Self::create_lottery(
+            env,
+            ticket_price,
+            number_of_numbers,
+            max_range,
+            thresholds,
+            min_players_count,
+        );
+    }
+
+    /// Creates the new lottery.
+    /// This function must be invoked by the administrator.
+    /// It can be called each time after previous lottery
+    /// has been completed. New lottery can have different specs,
+    /// pool balance is carried over from the previous one.
+    /// All previously stored tickets are cleared
+    ///
+    /// # Arguments
+    ///
+    /// - `env` - The environment for this contract.
+    /// - `ticket_price` - Unitary ticket price for the current lottery.
+    /// - `number_of_numbers` - Number of numbers possible to select by players
+    /// - `max_range` - Right boundary of the range players will select numbers from (1, max_range)
+    /// - `thresholds` - Thresholds with prizes for correctly selected numbers (specified as percentage of the pool balance)
+    /// - `min_players_count` - Minimum number of players needed to play the lottery
+    pub fn create_lottery(
```

Comment:
This should not be pub as it is not callable without init.

Reply:
It is actually. The idea is to create a new lottery each time the previous one ends. Each new lottery can have different rules. That way unspent tokens can be carried over to new lottery and increase its pool.